snuba/snuba_migrations/events_analytics_platform/0029_smart_autocomplete_items.py

```diff
@@ -0,0 +1,159 @@
+from typing import List, Sequence
+
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.datasets.storages.tags_hash_map import get_array_vals_hash
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.migrations.operations import OperationTarget, SqlOperation
+from snuba.utils.schemas import Array, Column, Date, String, UInt
+
+num_attr_buckets = 40
+
+columns: List[Column[Modifiers]] = [
+    Column("organization_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("item_type", UInt(8)),
+    Column("date", Date(Modifiers(codecs=["DoubleDelta", "ZSTD(1)"]))),
+    Column(
+        "retention_days",
+        UInt(16),
+    ),
+    Column(
+        "attributes_string_hash",
+        Array(
+            UInt(64), Modifiers(materialized=get_array_vals_hash("attributes_string"))
+        ),
+    ),
+    Column("attributes_string", Array(String())),
+    Column(
+        "attributes_float_hash",
+        Array(
+            UInt(64), Modifiers(materialized=get_array_vals_hash("attributes_float"))
+        ),
+    ),
+    Column("attributes_float", Array(String())),
+    # a hash of all the attribute keys of the item in sorted order
+    # this lets us deduplicate rows with merges
+    Column(
+        "key_hash",
+        UInt(
+            64,
+            Modifiers(
+                materialized="cityHash64(arraySort(arrayConcat(attributes_string, attributes_float)))"
+            ),
+        ),
+    ),
+]
+
+
+_attr_num_names = ", ".join(
+    [f"mapKeys(attributes_float_{i})" for i in range(num_attr_buckets)]
+)
+_attr_str_names = ", ".join(
+    [f"mapKeys(attributes_string_{i})" for i in range(num_attr_buckets)]
+)
+
+
+MV_QUERY = f"""
+SELECT
+    organization_id AS organization_id,
+    project_id AS project_id,
+    item_type as item_type,
+    toDate(timestamp) AS date,
+    retention_days as retention_days,
+    arrayConcat({_attr_str_names}) AS attributes_string,
+    arrayConcat({_attr_num_names}) AS attributes_float
+FROM eap_items_1_local
+"""
+
+
+class Migration(migration.ClickhouseNodeMigration):
+
+    blocking = False
+    storage_set_key = StorageSetKey.EVENTS_ANALYTICS_PLATFORM
+    granularity = "8192"
+
+    local_table_name = "eap_item_attrs_1_local"
+    dist_table_name = "eap_item_attrs_1_dist"
+    mv_name = "eap_item_attrs_1_mv"
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        create_table_ops = [
+            operations.CreateTable(
+                storage_set=self.storage_set_key,
+                table_name=self.local_table_name,
+                engine=table_engines.ReplacingMergeTree(
+                    storage_set=self.storage_set_key,
+                    primary_key="(organization_id, project_id, date, key_hash)",
+                    order_by="(organization_id, project_id, date, key_hash)",
```

Comment:
Should we keep the retention days in the sort key to deduplicate between 2 identical rows with short and long retention period? I wouldn't want the value to vanish when it's actually a valid value for a span stored for a longer period.

Reply:
Done